90zfsbootmenu/zfsbootmenu-lib.sh

```diff
@@ -2120,6 +2121,65 @@ zbmcmdline() {
   [ -f "${BASE}/zbm.cmdline" ] && echo | cat "${BASE}/zbm.cmdline" -
 }
 
+# arg1: pid
+# prints: child pid
+# returns: 0 if a child pid was found, 1 if there are no children
+
+find_child_pid() {
+  local pid child
+
+  pid="${1}"
+  if [ -z "${pid}" ]; then
+    zdebug "empty pid"
+    return 1
+  fi
+
+  if [ -e "/proc/${pid}/task/${pid}/children" ] ; then
+    read -r child < "/proc/${pid}/task/${pid}/children"
+
+    if [ -n "${child}" ] ; then
+      echo "${child}"
+      return 0
+    fi
+  fi
+
+  return 1
+}
+
+# prints: nothing
+# returns: nothing
+
+takeover() {
+  local pid child
+
+  # Kill the other running instance
+  if [ -e "${BASE}/active" ] ; then
+    read -r pid < "${BASE}/active"
+    parent=${pid}
+
+    echo "Stopping active zfsbootmenu with a PID of ${pid}"
+
+    # Trip the USR1 handler in /bin/zfsbootmenu - 'exit 0'
+    zdebug "sending USR1 to ${parent}"
+    kill -USR1 "${parent}"
+
+    # find the last child process of the active /bin/zfsbootmenu
+    while child="$( find_child_pid "${parent}" )" ; do
+      if [ -n "${child}" ] ; then
+        parent=${child}
+        continue
+      fi
+      break
+    done
+
+    # Kill the blocking child so that the USR1 handler actually executes
+    zdebug "killing child process ${parent}"
+    [ -n "${parent}" ] && kill "${parent}"
```

Comment:
I'm out of my depth here, but it would be great if there were built-in job control primitives that would do this without requiring a tree walk.

Reply:
I looked and nothing jumped out at me. I'll dig around a bit more though, this loop and the logic in `find_child_pid` feels sketchy.

Reply:
The only other way that I can see to do this is to use `ps --sid <session id> -o pid -h` to find all of the processes in a fixed session ID, but there's a lot of faffing about parsing the output of `ps` to discover the session ID of the active zfsbootmenu process, then find all the pids in that session, etc. A /proc tree walk is probably the best interface here for the problem.

Reply:
Yeah; if we have to signal children manually, walking `/proc` is as clean as it gets.

We need bash to implement `set --interruptible-children` or something to just do the right thing without our involvement.